### ISSUE TITLE
Common superclass for Sketchup::ComponentInstance and Sketchup::Group

### ISSUE THREAD

**Issue by Aerilius** [SketchUp/api-issue-tracker#179]
As discussed among some developers on BaseCamp, the SketchUp could provide a common superclass for components and groups, like `Sketchup::Container`.

Use cases:
- Since both share so many properties that some plugin functions can support both types, we developers often write unnecessarily complicated type-checking or filtering code:
  - `if entity.is_a?(Sketchup::ComponentInstance) || entity.is_a?(Sketchup::Group)`
  - `entities.grep(Sketchup::ComponentInstance).concat(entities.grep(Sketchup::Group))`
  - `entity.respond_to?(:definition)`
  - Yardoc: `@param containers [Array<Sketchup::ComponentInstance, Sketchup::Group>]`
  Having a common superclass would simplify code that handles entities that only need any of the shared properties of components and groups.

Other reasons:
- Components and groups share a significant amount of properties that we assume they already internally derive from a common superclass. Actually we discovered all of us don't care whether an entity is a group or component.
  - metadata: `name`, `description`, `guid`
  - definition: `definition` including inherited behaviors (snap_to, always_face_camera)
  - transformable: `transformation`
  - containing entities: `definition.entities` (where Group subclass has shortcut `group.entities`), `explode`
  - solidifyable: `volume`, `manifold?`, `intersect`, other solid operations
  - gluable: `glued_to`

**Comment by thomthom** [comment]
Internally Groups are sub-classes of ComponentInstance.
We can't change that now without breaking things, but we could introduce a new subclass, `Instance` for example that would allow us to write more generic code in the future.

**Comment by Aerilius** [comment]
If `Sketchup::Group` becomes a subclass of `Sketchup::ComponentInstance`, current code `obj.is_a?(Sketchup::ComponentInstance)` (expecting not a group) would then return positive for both and break.

- And what if you just add the common super class (or mixin module) only on the Ruby side, like the implicit tool or observer superclasses?
- What if you see it as an opportunity to refactor the internal code, in case ComponentInstance makes more sense as a subclass of a common superclass?
- And otherwise I'll wait curiously for Ruby 3 ["soft-typing or something like that"](https://blog.heroku.com/ruby-3-by-3/) or "type interfaces"...

**Comment by Eneroth3** [comment]
I've been thinking about writing this request myself for some time, but don't know how likely Trimble is to implement is.

A shared superclass Container for Group and ComponentInstance, and another one, Instance, also including Image, would make our Ruby code so much more beautiful.

The actual behaviors can still belong to the current classes if that makes it easier to implement, and empty classes could be inherited or mixed in.

**Comment by DanRathbun** [comment]
Try a test ...
```ruby
Sketchup::Container = Module.new # an empty mixin module

Sketchup::ComponentInstance.class_eval { include Sketchup::Container }
Sketchup::Group.class_eval { include Sketchup::Container }
```

Then see what the `#ancestors(0` method returns ...
```ruby
Sketchup::ComponentInstance.ancestors
#=> [Sketchup::ComponentInstance, Sketchup::Container, Sketchup::Drawingelement, Sketchup::Entity,
#=> Object, JSON::Ext::Generator::GeneratorMethods::Object, Kernel, BasicObject]
Sketchup::Group.ancestors
#=> [Sketchup::Group, Sketchup::Container, Sketchup::Drawingelement, 
#=> Sketchup::Entity, Object, JSON::Ext::Generator::GeneratorMethods::Object, Kernel, BasicObject]
```

Then test `#is_a?()` ...
```ruby
obj = Sketchup.active_model.selection[0] # Stacy instance selected
obj.is_a?(Sketchup::Container)
#=> true
```

My only wonder is would "`Sketchup::GeomContainer`" or "`Sketchup::GeometryContainer`" be a better (more descriptive) superclass name ?

**Comment by Eneroth3** [comment]
Geom is used for "abstract" geometry (points, lines planes) so I think it could be misleading here. EntitiesContainer would be better but I think just Container is enough.

**Comment by kengey** [comment]
I believe class inheritance is a bad thing since things cannot be lously coupled by doing so. This breaks the SOLID open/closed design principle: [https://en.wikipedia.org/wiki/SOLID](link). And, eventually, as a developer, you are not really interested in what type of element is passed (except for valuetypes), but what it implements, thus what methods and functions the object exposes. 

I could imagine having a constant array of methods for all "interfaces" and having an InterfaceChecker that checks if an object has all methods we are interested in.

**Comment by Aerilius** [comment]
I believe this is very opinionated, and I didn't find any proofs on the internet, although inheritence clearly has also disadvantages. The focus here is not on inheriting code (which is internal to SketchUp anyways), but marking types that they share behavior. That is what interfaces or traits are for, but Ruby knows neither of them.

Comparing a list of methods would not be enough, the complete method signatures must match (parameter and return types, and ideally the intention what the method does). But this is up to the language to implement not SketchUp.

---

So still we have the option to leave it as it is, and 
- check for multiple types: `entity.is_a?(Sketchup::ComponentInstance) || entity.is_a?(Sketchup::Group)` 
  although as a developer I am interested in the "**Containable trait**" that they **contain** something (and have definition.entities and related methods)
- or do duck-typing and check for the presence of every used method name `entity.respond_to?(:definition)`
  but that is in my opinion too wide, and does not exclude objects that by chance also implements methods of the same name with different behavior or interface.

**Comment by kengey** [comment]
I agree this is opinionated,  just like every pattern might be considered an antipattern for someone else. However, I still believe the essence is to know what the object exposes, and not what it actually is. Consider your object is exploding things, all it then cares is if the argument is implementing the explode method, with the correct number of arguments. Consider that same code outside of sketchup but also handling the same logic on other types that implement the explode function.

In my opinion. No class should ever be aware of another type. Dependencies should be injected (see dependency injection) and objects should be handled depending on what they can, and not depending on what they are.

**Comment by DanRathbun** [comment]
FTR, the example I showed above is composition and not inheritance, which is often preferable for many reasons.

**Comment by DanRathbun** [comment]
**EDIT**: After a year, I revisited this topic. I've changed my mind on the name "Container" since groups and instances do not themselves "own" a container of geometry (**`entities`** collection).
It is their definitions that "own" the geometry containers. 

It seems about weekly that I have to explain to new API coders that groups and instances do not actually have their own geometric entities collection. I feel using the ancestor mixin "Container" would serve to confuse new API coders.

So I like instead Thomas' suggestion for a the name **`Sketchup::Instance`**.

Also since we do not want actual instances of the **`Sketchup::Instance`** objects, it should be a mixin module rather than a class.

So, I've edited the test example [above](https://github.com/SketchUp/api-issue-tracker/issues/179#issuecomment-440322662) to reflect this.

----

SIDENOTE: Some readers may also recall that I've politicized for a new `Sketchup::Container` class (perhaps a subclass of `Sketchup::Group`,) that would only allow 1 instance, which can only be added to the model or another `Sketchup::Container` instance. (This would mean, like groups they are not shown in the Components inspector, ... and that any imports with containers must be imported to the toplevel model or an existing container object. And yes they would appear in the Outliner as objects.)
